crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -53,6 +58,8 @@ pub struct Room {
     own_user_id: Arc<UserId>,
     inner: Arc<SyncRwLock<RoomInfo>>,
     store: Arc<dyn StateStore>,
+    forward_timeline_streams: Arc<Mutex<Vec<mpsc::UnboundedSender<TimelineSlice>>>>,
+    backward_timeline_streams: Arc<Mutex<Vec<mpsc::UnboundedSender<TimelineSlice>>>>,
```

Comment:
> Could you elaborate why this needs to be Unbounded-Channels (rather than bounded ones) specifically?

I used the an Unbound-Channel since there isn't any real limit.

How would a unbounded channel leak memory? Isn't the channel closed when the last sender or receiver is dropped and then all messages are dropped as well.

Reply:
> How would a unbounded channel leak memory? 

Anything without a specific limit that things are just pushed to without regard is always a cause for concern in growing too much memory. Also any hashmap that acts as a cache, for e.g., will just grow over its lifetime and isn't cleaned up. From the outside, you only notice the process claiming and not releasing more and more memory until it eventually runs out. This doesn't qualify as a c-style-memory-leak in the classic sense (which is bad, because tools like valgrind don't detect it), but is generally cause for concern, in particular because these are super hard to track down (tools don't detect them, as it is _validly_ claimed memory). I had to track down a good amount of these and am very wary of introducing anything that can contentiously grow without stop.

I've come to learn that in practice there is  _rarely_ good reason for an unbounded channel, but it is just easy to use and implement with bad outcomes down the line. I have previously also raised that problem with the tokio team, because for most users the bad outcomes from this API are hard to understand, but they didn't consider it their problem or cause for change.

>  Isn't the channel closed when the last sender or receiver is dropped and then all messages are dropped as well.

Yes, of course, this only happens on "wrong usage". The problem with unbounded channels (in this case wrapped as a stream) is that it is easier to use them wrong than it is to make sure they are used right. And while the _wrong usage_ happens elsewhere, the unlimited growing happens from here, so by the extend to the user of the API it is _our_ responsibility (rather than theirs to continuously read the stream). I can think of variety of scenarios, where this could occur. One you accidentally created yourself by defining the streams with a `&'static` lifetime - this would mean they are _never_ dropped and when not consumed would grow to unpractical sizes. Similarly I could think of a scenario where a screen is using the fetch method for a room and while the user is navigating elsewhere is not disregarded properly, keeping the reference alive - yet when they come back a new screen is created, creating a new channel. While the _bug_ itself is on the opposite side of the implementation, we would keep feeding these without limitation and only be visible in weird memory growth patterns over time - rather than, e.g. explicit warn-messages when we have to start dropping because the queue is full.

> I used the an Unbound-Channel since there isn't any real limit.

There should be. The limit of a channel is only for messages _in transit_ within the queue - it limits the memory size the channel can grow to. As a user of the API you should not be holding the queue if you are not consuming it regularly. It is better to discard and recreate the queue, maybe we should make that explicit in the docs for it. But either case, what _valid_ use case is there for having a channel of timeslices grow beyond, say, 5 items? Why should we be supporting this pattern at all?

Instead let's limit the channel size, put that limit in the docs and put out a fat warning that someone is not consuming their stream (and drop the sender) when we encounter it.

Reply:
Thanks for explaining. I will switch to bounded channels